third_party_plugin_dev_portal/VST3presetSwitchingPerformEditAlternatives.md

```diff
@@ -0,0 +1,199 @@
+
+[https://forums.steinberg.net/t/how-to-use-restartcomponent-and-which-flags-are-the-right-one-when-changing-all-characteristics-parameters-except-size/202031/13](https://forums.steinberg.net/t/how-to-use-restartcomponent-and-which-flags-are-the-right-one-when-changing-all-characteristics-parameters-except-size/202031/13)
+
+> If you now call restartComponent with `kParamTitlesChanged|kParamValuesChanged`, the host should have the new parameter names, flags and values. The host will not send the values to your processor in the process call this way. So if you need them there, but without going thru the beginEdit, performEdit, endEdit methods, you could allocate an IMessage with `IHostApplication::createInstance()` fill it with the necessary information and send it via `IConnectionPoint::notify` to your processor.
+
+## Relevant VST3 documentation
+
+### Optional separation of processor and edit controller
+
+[https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#basic-concept](https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#basic-concept)
+
+> The design of VST 3 suggests a complete separation of processor and edit controller by implementing two components. […] A plug-in that supports this separation has to set the Steinberg::Vst::kDistributable flag in the class info of the processor component (Steinberg::PClassInfo2::classFlags). Of course not every plug-in can support this, for example if it depends deeply on resources that cannot be moved easily to another computer. So when this flag is not set, the host must not try to separate the components in any way.
+
+### private communication (custom IMessage)
+
+[https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#private-communication](https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/API+Documentation/Index.html#private-communication)
+
+Data that is unknown to the host can be transmitted by means of messages. The communication interfaces are
+* Steinberg::Vst::IConnectionPoint: The host establishes a connection between processor and controller.
+* Steinberg::Vst::IMessage: Represents a message to send to the counterpart.
+* Steinberg::Vst::IAttributeList: A list of attributes belonging to a message.
+
+Please note that messages from the processor to the controller must not be sent during the process call, as this would not be fast enough and would break the real time processing. Such tasks should be handled in a separate timer thread.
+
+## Implementation hints from open source frameworks
```

Comment:
Not sure we should include this section at all to be honest.

Reply:
because frameworks are constantly changing... they should all implement the VST3 API as correctly as possible so providing a good example with the VST3 SDK is enough imo.

Reply:
> frameworks are constantly changing = more possibility of out of date info on our docs pages